// @mysten/suins/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  ALLOWED_METADATA: () => import_constants.ALLOWED_METADATA,
  SuinsClient: () => import_suins_client.SuinsClient,
  SuinsTransaction: () => import_suins_transaction.SuinsTransaction,
  getCoinDiscountConfigType: () => import_helpers.getCoinDiscountConfigType,
  getConfigType: () => import_helpers.getConfigType,
  getDomainType: () => import_helpers.getDomainType,
  getPricelistConfigType: () => import_helpers.getPricelistConfigType,
  getRenewalPricelistConfigType: () => import_helpers.getRenewalPricelistConfigType,
  isNestedSubName: () => import_helpers.isNestedSubName,
  isSubName: () => import_helpers.isSubName,
  mainPackage: () => import_constants.mainPackage,
  validateYears: () => import_helpers.validateYears
});
module.exports = __toCommonJS(index_exports);
var import_suins_client = require("./suins-client.js");
var import_suins_transaction = require("./suins-transaction.js");
var import_constants = require("./constants.js");
var import_helpers = require("./helpers.js");
