// @mysten/suins/dist/cjs/pyth/PriceServiceConnection.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PriceServiceConnection_exports = {};
__export(PriceServiceConnection_exports, {
  PriceServiceConnection: () => PriceServiceConnection
});
module.exports = __toCommonJS(PriceServiceConnection_exports);
var import_axios = __toESM(require("axios"));
var import_axios_retry = __toESM(require("axios-retry"));
class PriceServiceConnection {
  /**
   * Constructs a new Connection.
   *
   * @param endpoint endpoint URL to the price service.
   * @param config Optional configuration for custom setups.
   */
  constructor(endpoint, config) {
    this.httpClient = import_axios.default.create({
      baseURL: endpoint,
      timeout: config?.timeout || 5e3
    });
    (0, import_axios_retry.default)(this.httpClient, {
      retries: config?.httpRetries || 3,
      retryDelay: import_axios_retry.default.exponentialDelay
    });
  }
  /**
   * Fetch latest VAAs of given price IDs.
   *
   * @param priceIds Array of hex-encoded price IDs.
   * @returns Array of base64 encoded VAAs.
   */
  async getLatestVaas(priceIds) {
    const response = await this.httpClient.get("/api/latest_vaas", {
      params: {
        ids: priceIds
      }
    });
    return response.data;
  }
}
